### PR TITLE
Added optional argument support to the schema parser

### LINKED ISSUE #20113: [JIT][script] JIT for torch.norm ignores argument defaults
## 🐛 Bug

torch.norm `p` argument default is ignored by JIT

## To Reproduce

1.
```python
import torch

@torch.jit.script
def norm_test():
  t = torch.ones(10, 5)
  return torch.norm(t, dim=1, keepdim=True)
```

fails with:

> RuntimeError:
> arguments for call are not valid:
>   
>   for operator aten::norm(Tensor self, Scalar p=<default>) -> Tensor:
>   keyword argument dim unknown
>   
>   for operator aten::norm(Tensor self, Scalar? p, *, int dtype) -> Tensor:
>   argument p not provided.
>   @torch.jit.script
>   def norm_test():
>     t = torch.ones(10, 5)
>     return torch.norm(t, dim=1, keepdim=True)
>            ~~~~~~~~~~ <--- HERE

2.
```python
import torch

@torch.jit.script
def norm_test():
  t = torch.ones(10, 5)
  return torch.norm(t, p=2, dim=1)
```
Works fine

## Environment

Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
CMake version: version 3.9.0-rc5

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 419.67
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\bin\cudnn64_7.dll

Versions of relevant libraries:
[pip] numpy==1.14.3
[pip] numpydoc==0.8.0
[pip] pytorch-ignite==0.1.0
[pip] torch==1.1.0
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl
[conda] cuda91                    1.0                           0    pytorch
[conda] cuda92                    1.0                           0    pytorch
[conda] mkl                       2018.0.2                      1
[conda] mkl-service               1.1.2            py36h57e144c_4
[conda] mkl_fft                   1.0.1            py36h452e1ab_0
[conda] mkl_random                1.0.1            py36h9258bd6_0
[conda] pytorch                   1.1.0           py3.6_cuda90_cudnn7_1    pytorch
[conda] pytorch-ignite            0.1.0                    pypi_0    pypi
[conda] torchvision               0.2.1                    pypi_0    pypi

cc @suo

### PR DESCRIPTION
**Pull request by Chillee**
Definitely one of those "figure out what's going on for hours and fix in a minute" things :/

Fixes https://github.com/pytorch/pytorch/issues/20113

### CODE REVIEW

File: torch/csrc/jit/script/schema_matching.cpp [hunk 1]
```diff
@@ -332,6 +332,8 @@ c10::optional<MatchedSchema> tryMatchSchema(
       // Argument has a default value and no value was provided, so use the
       // default
       actual_named_value = NamedValue(*arg.default_value());
+    } else if (arg.type()->cast<OptionalType>()) {
+      actual_named_value = NamedValue(IValue());
```
Comment (zou3519) [inline torch/csrc/jit/script/schema_matching.cpp]: Add test?

Comment (apaszke) [reply torch/csrc/jit/script/schema_matching.cpp]: Doesn't this mean that optionals always have default values? It's not something that we would want.

Comment (Chillee) [reply torch/csrc/jit/script/schema_matching.cpp]: Added a test, not entirely sure if it's the right place.

File: torch/csrc/jit/script/schema_matching.cpp [hunk 2]
```diff
@@ -332,6 +332,8 @@ c10::optional<MatchedSchema> tryMatchSchema(
       // Argument has a default value and no value was provided, so use the
       // default
       actual_named_value = NamedValue(*arg.default_value());
+    } else if (arg.type()->cast<OptionalType>()) {
```
Comment (driazati) [inline torch/csrc/jit/script/schema_matching.cpp]: It seems like this is saying that any optional arguments would implicitly have `None` as a default value? We should be pulling default arguments from the function schema's themselves instead of inserting it automatically.

Comment (wanchaol) [reply torch/csrc/jit/script/schema_matching.cpp]: Yeah that's also what I suggested. This should be the schema's responsibility to ensure a None defaults

File: test/test_jit.py [hunk 3]
```diff
@@ -748,6 +748,14 @@ def fn(*args):
                 self.assertFalse(fn.has_trace_for(*unk_config))
         self.assertEqual(fn.hits, 0)
 
+    def test_optional_args(self):
+        @torch.jit.script
+        def norm_test():
```
Comment (wanchaol) [inline test/test_jit.py]: Instead of adding it here, can you add a entry to https://github.com/pytorch/pytorch/blob/master/test/common_methods_invocations.py#L524? JIT will automatically pick tests there

File: torch/csrc/jit/script/schema_matching.cpp (hunk 2 above)
Comment (Chillee) [reply torch/csrc/jit/script/schema_matching.cpp]: Are you guys saying that when we wrote the schema we should have made sure that all optional arguments had a default argument?

Comment (wanchaol) [reply torch/csrc/jit/script/schema_matching.cpp]: Yeah i think so. Thinking of it more, I think the real problem here is that sometimes schema cannot specify the default argument for middle args because C++ cannot have default arg before non-default arg, like `p` in `norm`. Maybe we should do this in schema matching then?

@driazati what do you think?

Comment (driazati) [reply torch/csrc/jit/script/schema_matching.cpp]: The other args are also missing defaults, at least according to [the docs](https://pytorch.org/docs/stable/torch.html#torch.norm), so it's still at the end and a valid Python schema.

Comment (wanchaol) [reply torch/csrc/jit/script/schema_matching.cpp]: sounds good to me. that indeed means schema in native_functions.yaml need to unify to avoid overloads ambiguity for the `norm` and `empty_like` case.

**Comment by Chillee** [comment]
I tried to follow the path that @wanchaol and @driazati agreed upon, but I gave up. For those braver than me, this was the path I went on before I gave up:
```
I add default arguments to scalar? p -> invalid generated C++ code since we have named arguments before positional arguments -> I consolidate everything down to one function with a bunch of optional arguments -> IntArrayRef is not a valid type for gen_jit_dispatch -> I add some stuff in -> name collision error for norm_out -> ???
```
